### PR TITLE
Add `Channel#try_send`, a non-blocking `Channel#send`

### PR DESCRIPTION
**Pull request by carlhoerberg**
Instead of having to write
```crystal
select
when @channel.send 1 then true
else false
end
```
and paying the `to_a` price in `select_impl` when wanting to send to a `Channel` non blocking.

In one of our apps we use this pattern a lot: 

```crystal
@flow = true
@flow_change = Channel(Bool).new

def wait_for_flow
  return if @flow
  until @flow_change.receive
  end
end

def flow=(val)
  @flow = val
  while @flow_change.try_send(val) # notify all flow change subscribers
  end
end
```

So the `flow_change` channel may or may not have receivers.

### CODE REVIEW

**Comment by straight-shoota** [comment]
I think it might be good to add this behavioral variant.

But the name `#send?` is confusing. It would suggest similar semantics as `#receive?`, which returns `nil` instead of raising `ClosedError`. But that's something different. This method still raises `ClosedError`. The only difference to `#send` is that it returns instead of waiting when immediate delivery is impossible.
Not sure what would be a better name, tho. `#try_send?`? 🤔

**Comment by asterite** [comment]
I think instead we should change the code to avoid the to_a call. For example there's no need to call to_a if the Indexable has a single element.

**Comment by asterite** [comment]
I tried to optimize this a bit, but the main issue is that even with the optimization the code is translated to create a `SendAction`, which is a class so it allocates memory.

I think the right thing to do is, yes, to introduce some form of `send?`, but also change the expansion of `select ...` with just one channel.send to use `send?` instead.

In my mind `select` should be the only primitive to use (like in Go) and it's the compiler's responsibility to optimize it for the single case, not the user's responsibility.

**Comment by asterite** [comment]
I'm also surprised to learn that a `select` expression creates a class for each of the actions (like `SendAction`) so memory is allocated on every `select`. Really bad! I think it should be done in a way that no memory is ever allocated. Like, probably `SendAction` should be a struct, and we create a static array to hold those values in the stack. Then use pointers to refer to those. I think Go might be doing the same thing.

Then we can optimize for the case of 1 select action. In the end it will be as efficient as `send?`, I think.

It's more work for the compiler/us, but it's work that only needs to be done once by us, and not put the burden on developers.

**Comment by asterite** [comment]
Also it seems an Indexable is passed to select_impl. In the actual implementation that's a tuple. But that's not good if we need to sort the values. So I would change the implementation to use a Static Array and pass a slice to the method. Then the slice is mutable so the data can be sorted in-place. And maybe then the data can also be structs to avoid memory allocations. But I see some pointers being passed around, so it's a tricky refactor. But even though it'll take some time to get it right, it's probably the best thing to do.

**Comment by yxhuvud** [comment]
> In my mind select should be the only primitive to use

Considering that a very common use case is to wrap the select in a loop, I could see a second variant of `select` that comes with the loop built in and some predefined exit behavior semantics. That would allow for optimizations that is not really reasonable with select as it looks today. For example if you are selecting off several different things it would allow you to not tear down and then set up the waiting again each iteration of the loop - it would suffice to redo the one that triggered.

File: src/channel.cr [hunk 1]
```diff
@@ -246,6 +246,45 @@ class Channel(T)
...
+  # Returns `true` if the value could be deliviered, `false` if not or if the channel is closed.
+  def try_send?(value : T) : Bool
+    @lock.lock
+    case send_internal(value)
+    in .delivered?
+      true
+    in .closed?
+      false
+    in .none?
+      false
```
Comment (Sija) [inline src/channel.cr:282]: Could be shortened a bit:

```suggestion
    in .closed?, .none?
      false
```

File: spec/std/channel_spec.cr [hunk 2]
```diff
@@ -889,4 +889,31 @@ describe "buffered" do
     ch = Channel(Int32).new(10)
     ch.pretty_inspect.should eq("#<Channel(Int32):0x#{ch.object_id.to_s(16)}>")
   end
+
+  it "can send non blocking" do
+    ch = Channel(Int32).new
+    spawn(same_thread: true) do
+      ch.receive.should eq 1
+    end
+    Fiber.yield # required to let the receiving fiber start
```
Comment (straight-shoota) [inline spec/std/channel_spec.cr:898]: I'd suggest to use the channel for synchronization between fibers. We already have it anyways and that's more robust than depending on `Fiber.yield` and `same_thread: true` to run the spawned fiber at that point.

```suggestion
    spawn do
      ch.send 0 # sync
      ch.receive.should eq 1
    end
    ch.receive # let the receiving fiber start
```

File: spec/std/channel_spec.cr [hunk 3]
```diff
@@ -889,4 +889,31 @@ describe "buffered" do
...
+    ch.try_send(1).should be_true
+    ch.try_send(2).should be_false
+  end
+
+  it "can send non blocking to buffered channel" do
+    ch = Channel(Int32).new(1)
+    spawn(same_thread: true) do
+      ch.receive.should eq 1
+    end
+    Fiber.yield # required to let the receiving fiber start
```
Comment (straight-shoota) [inline spec/std/channel_spec.cr:908]:
```suggestion
    spawn do
      ch.send 0 # sync
      ch.receive.should eq 1
    end
    ch.receive # let the receiving fiber start
```

**Comment by beta-ziliani** [comment]
Hi @carlhoerberg , we'd like to move forward with this. What do you think of straight-shoota's observations? Do you want us to take it to completion? Thanks!

**Comment by asterite** [comment]
> we'd like to move forward with this

Really? Could you show an equivalent API for Go? And if Go doesn't have one, why do we need one? I think it might be better to fix `select`. I know it's harder, but it's the right thing to do.

**Comment by carlhoerberg** [comment]
Isnt this less verbose than any select, fixed or not?

**Comment by asterite** [comment]
I guess that's right. My only concern is that eventually if we fix `select` to be as fast as `try_send` then there will be two ways to do the same thing. Maybe it's not a big deal. We could also eventually rewrite `select` to `try_send` if we detect they are the same.

**Comment by bcardiff** [comment]
The implementation seems correct, but I worry about increasing Channel's API. It's easy for things go wrong and not having a single code for sending things through the channel seems like a bad idea. In this case the whole method is blocking so it should be safe, but in general things like https://github.com/crystal-lang/crystal/blob/1b932187607adf7abb41968c8a36e62e2c2dbebd/src/channel.cr#L453-L457 makes me shiver.

We have `receive`/`receive?` which are both blocking, having `try_send` to be non-blocking seems confusing and opens the question whether we are going to add `try_receive`/`send?` for consistency. IMO it will not be trivial to remember that `?` is blocking but `try_` is not.

I also wonder if people would expect to write `select; when ch1.try_send ...; when ch2.try_send ...;` which will not compile because there is no matching `_action` method.

If `select;when ch.send x then true; else false; end` is a common pattern, I would prefer to put that in a macro or method and optimize the `select_impl` implementation a bit.

From what I see, it seems that the culprit of the `to_a` is only: https://github.com/crystal-lang/crystal/blob/1b932187607adf7abb41968c8a36e62e2c2dbebd/src/channel.cr#L426-L431

Did someone check if doing a special case to use a `Slice` to point to a single element when `ops.size == 1` is enough to avoid the extra allocations?

If that is not enough, in order to keep the API as today we could have a `private def self.select_impl_single(op : SelectAction, non_blocking)`, with more optimizations that can be applied for the single channel case, even at the cost of some internal code duplication that can be solved/refactored later.

A bonus of either last approaches is that will also apply to `select; when ch.receive; else ...` directly.

**Comment by straight-shoota** [comment]
@bcardiff Sounds like a good assessment. 👍 

Optimizing `#select_impl` should happen anyways. For a single action (or even a small number) we can avoid an array allocation. Additionally, we can try to refactor `SendAction` without heap allocations as suggested in https://github.com/crystal-lang/crystal/pull/12694#issuecomment-1296236424

Thus the lengthy form with a single `select` branch should be as performant as the custom implementation of `#try_send`.
I suppose at that point we could technically even implement `#try_send` based on that so that all Channel interactions go through a single point.
But then we could also consider other alternatives for this pattern in order to keep the Channel API succinct and avoid potential confusion about `#try_send`.

**Comment by bcardiff** [comment]
I was thinking that something like https://github.com/crystal-lang/crystal/compare/master...bcardiff:crystal:select-optimizations would avoid the array heap allocation, and can will cover single and dual action selects. WDYT?

Essentially it applies the following logic to keep a tuple of ops when possible for the locks.

```crystal
def f(*ops)
  f_impl ops
end

def f(ops : Indexable)
  f_impl ops
end

def f_impl(ops : Indexable)
    if ops.is_a?(Tuple) && ops.size == 1
      f_impl_with_locks(ops, ops)
    elsif ops.is_a?(Tuple) && ops.size == 2
      ops0 = ops.fetch(0, nil).not_nil!
      ops1 = ops.fetch(1, nil).not_nil!

      case (ops0 <=> ops1)
      when 0
        f_impl_with_locks(ops, {ops0})
      when 1
        f_impl_with_locks(ops, {ops1, ops0})
      when -1
        f_impl_with_locks(ops, {ops0, ops1})
      else
        raise "unreachable"
      end
    else
      ops_locks = ops
        .to_a
        .uniq!
        .sort!

      f_impl_with_locks(ops, ops_locks)
    end
end

def f_impl_with_locks(ops : Indexable, ops_locks)
  puts({typeof(ops), typeof(ops_locks)})
end

f 1         # => {Tuple(Int32)              , Tuple(Int32)}
f 1, 2      # => {Tuple(Int32, Int32)       , Tuple(Int32, Int32)}
f 1, 2, 3   # => {Tuple(Int32, Int32, Int32), Array(Int32)}
f [1, 2]    # => {Array(Int32)              , Array(Int32)}
```

The runtime operations for the non blocking send is basically the proposal of this PR, so I expect it to work as efficient.

**Comment by straight-shoota** [comment]
I think it's easier than that. We can just turn the tuple into a `StaticArray` for sorting. There's no need for special treatment for individual cases. This improves performance for all sizes of `ops`, as long as the size is known at compile time. 

I have a patch for that ready that goes on top of #12814. It would work without it, but I'm batching it avoid conflicts.

On top of master the simplified diff would be this:

```diff
@@ -423,8 +423,12 @@ class Channel(T)
   private def self.select_impl(ops : Indexable(SelectAction), non_blocking)
     # Sort the operations by the channel they contain
     # This is to avoid deadlocks between concurrent `select` calls
-    ops_locks = ops
-      .to_a
+    if ops.is_a?(Tuple)
+      ops_locks = ops.to_static_array
+    else
+      ops_locks = ops.to_a
+    end
+    ops_locks
       .uniq!(&.lock_object_id)
       .sort_by!(&.lock_object_id)
```

The implementation of `Tuple#to_static_array` would be analog to `StaticArray#[]`, returning `StaticArray(Union(*T), {{ T.size}})`. This can also be an internal helper method, but IMO it would be useful to have this publicly available. Doesn't matter though for `select_impl`.

**Comment by bcardiff** [comment]
Something I wanted to avoid is to have a union in ops_locks, so there is no multi-dispatch on each iteration.
Another thing that worries me is that `Tuple#to_static_array` will expand multiple defs. One for each combination of select actions used in the entire program. They should not be many but we are not forcing an upper bound unless we do something like I proposed.

---

> The runtime operations for the non blocking send is basically the proposal of this PR, so I expect it to work as efficient.

I was wrong, we still have the allocation of the `SelectContext` ary pointed out. But I would start with optimizing how locks are computed. Seems we agree there.

**Comment by straight-shoota** [comment]
Okay, yeah multi dispatch is indeed an issue. But we can't really prevent that on scale, unless we multiply all possible combinations.
I suppose we can combine both approaches and use the tuple sorting you suggested for small sizes.

But for the `else` branch I presume the `StaticArray` solution should be an improvement over `to_a`. For `T.size == 1` it probably doesn't make any difference whatsoever. So we're effectively optimizing for the case `T.size == 2`. Perhaps we could do the same for `T.size == 3`, but the tuple combinations just explode from there.

**Comment by straight-shoota** [comment]
Actually, your implementation doesn't avoid multi dispatch either, does it? The type of `ops.fetch(0, nil).not_nil!` is `Union(*T)`.

Consequently:
```cr
f 1      # => {Tuple(Int32)              , Tuple(Int32)}
f 1, 2u32      # => {Tuple(Int32, UInt32), Tuple(Int32 | UInt32, Int32 | UInt32)}
f 1, 2u32, 3   # => {Tuple(Int32, UInt32, Int32), Array(Int32 | UInt32)}
f [1, 2u32]    # => {Array(Int32 | UInt32), Array(Int32 | UInt32)}
```

`Tuple(Int32 | UInt32, Int32 | UInt32)` is not an improvement over `StaticArray(Int32 | UInt32, 2)` (one could argue that it's pretty much *the same thing*).

To avoid those union types, we need to switch over the tuple size at compile time. This should work with macros:
```cr
def f_impl(ops : Tuple(*T)) forall T
  {% if T.size == 1 %}
    f_impl_with_locks(ops, ops)
  {% elsif T.size == 2 %}
      case (ops[0] <=> ops[1])
      when 0
        f_impl_with_locks(ops, {ops[0]})
      when 1
        f_impl_with_locks(ops, {ops[1], ops[0]})
      when -1
        f_impl_with_locks(ops, ops)
      else
        raise "unreachable"
      end
  {% else %}
    f_impl(ops.to_a)
  {% end %}
end
```

**Comment by straight-shoota** [comment]
I suppose we could also look into alternative solutions to maybe avoid sorting altogether.

Considering that the size of `ops` is typically not that big, maybe iterating multiple times wouldn't be such a bad option.

**Comment by straight-shoota** [comment]
I don't follow that latest example. Where would those methods be called?

Looks like they are overloads of the splat variant of `non_blocking_select` which are actually unused and proposed for removal in #12813.

**Comment by bcardiff** [comment]
I thought the compiler would generate those calls. Either way doing a dispatch on runtime depending on the indexable size to a method that will avoid unions in the ops_locks can still be done. That's the core part of the idea.

**Comment by straight-shoota** [comment]
Yeah, sure that can be done. It's only feasible for very small sizes, though. Not even sure we would want to go higher than 2 or maybe 3.
And I wouldn't expect the actual performance improvement to be very noticeable. At least there are more effective measures, such as avoiding heap allocations.

**Comment by asterite** [comment]
I honestly wouldn't worry about multidispatch. We have LLVM here.

Here's an example:

```crystal
struct Int32
  def foo
    self
  end
end

class String
  def foo
    bytesize
  end
end

def sum(values : Union(Int32, String)[2])
  values.sum(&.foo)
end

fun my_awesome_function : Int32
  values = uninitialized Union(Int32, String)[2]
  values[0] = 1
  values[1] = "hello"

  sum(values)
end

puts my_awesome_function
```

We are initializing a static array whose type is a union. We put an int in the first position and a string in the second one. Then we "sum" the values, which would in theory result in multidispatch.

Compile the above program with `--release --emit llvm-ir` and search for `my_awesome_function` in the output `.ll` file. Here it is:

```llvm
define i32 @my_awesome_function() local_unnamed_addr #1 !dbg !25766 {
exit.1.i:
  ret i32 6, !dbg !25767
}
```

LLVM figured out what was in each position and completely avoided the multidispatch. Not only that, it computed everything at compile-time.

**Comment by asterite** [comment]
Unrelated to the comment above (I'll reply to it soon), look at this:

https://github.com/golang/go/blob/eaf0e3d4650fd223dec84ee52025c7a82bcb24bd/src/runtime/chan.go#L677-L715

So Go decides to implement different `select` patterns in different ways, probably optimal ones.

I think we could do the same. That is, instead of introducing `try_send`, detect this pattern and call a specific method (which would be `nodoc`.)

That way, at least for this issue, there's no need to optimize the general case.

**Pull request closed without merging**